spark3/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java

```diff
@@ -513,17 +512,22 @@ protected WriterFactory(PartitionSpec spec, FileFormat format, LocationProvider
 
     @Override
     public DataWriter<InternalRow> createWriter(int partitionId, long taskId, long epochId) {
-      OutputFileFactory fileFactory = new OutputFileFactory(
-          spec, format, locations, io.value(), encryptionManager.value(), partitionId, taskId);
-      SparkAppenderFactory appenderFactory = new SparkAppenderFactory(properties, writeSchema, dsSchema, spec);
+      Table table = tableBroadcast.getValue();
+
+      OutputFileFactory fileFactory = new OutputFileFactory(table, format, partitionId, taskId);
+      SparkAppenderFactory appenderFactory = new SparkAppenderFactory(table, writeSchema, dsSchema);
+
+      PartitionSpec spec = table.spec();
+      FileIO io = table.io();
+
       if (spec.isUnpartitioned()) {
-        return new Unpartitioned3Writer(spec, format, appenderFactory, fileFactory, io.value(), targetFileSize);
```

Comment:
I guess in future we can simplify these writer constructors too with table?

Reply:
I think so but it will have to touch more than one query engine so I stopped here for now. Sounds like a good follow-up item.

Reply:
Minor suggestion: maybe just refactor the one inside this file. Not touch the parent classes.

Reply:
Let me take care of that.

Reply:
Actually, I remember why I did not do this. If we were to expose writing data with a partition spec that is not current, then we would need to pass the correct partition spec here instead of consuming the default one from the table object. I think it should be alright to keep it as is for now.